src/yaml.cr

```diff
@@ -147,4 +148,28 @@ module YAML
   def self.dump(object, io : IO)
     object.to_yaml(io)
   end
+
+  # Determines if the scaler is reserved
+  def self.reserved_scalar(value : String)
```

Comment:
@straight-shoota I did EXACTLY what @ysbaddaden told me to do. Which didn't seem right either. @RX14, the YAML spec specifies some regexs to detect the scalar. How else would you do this?

Btw, here is the latest iteration using macros

```crystal
{% types = %w(Nil Bool Time Int8 Int16 Int32 Int64 UInt8 UInt16 UInt32 UInt64 Float32 Float64) %}
{% for type in types %}
  def {{type.id}}.new(pull : YAML::PullParser)
    location = pull.location
    begin
      case value = pull.read_plain_scalar
      {% if type == "Nil" %}
        when /^(~|null|Null|NULL)$/
          nil
      {% end %}
      {% if type == "Bool" %}
        when /^(y|Y|yes|Yes|YES|true|True|TRUE|on|On|ON)$/
          true
        when /^(n|N|no|No|NO|false|False|FALSE|off|Off|OFF)$/
          false
      {% end %}
      {% if type.starts_with?("Int") || type.starts_with?("Float") %}
        when /^[-+]?([0-9][0-9_]*)?\.[0-9.]*([eE][-+][0-9]+)?$/ # Base 10 int/float
          new(value, underscore: true)

        when /^[-+]?[0-9][0-9_]*(:[0-5]?[0-9])+\.[0-9_]*$/ # Base 60 int/float
          deg, min, sec = value.split(":")
            .map_with_index { |v| new(v, underscore: true, whitepsace: false) / 60 ** i }
            deg >= 0 ? deg + (min + sec) : deg - (min + sec)
      {% end %}
      {% if type.starts_with?("Int") %}
        when /^[-+]?0b[0-1_]$/ # base2 int
          new(value, base: 2, underscore: true)
        when /^[-+]?0[0-7_]+$/ # base8 int
          new(value, base: 8, underscore: true)
        when /^[-+]?0x[0-9a-fA-F_]+$/ # base16 int
          new(value, base: 16, underscore: true)
      {% end %}
      {% if type == "Time" %}
        when /^[0-9]{4}$/
          Time.parse(value, "%Y", Time::Kind::Utc)
        when /^[0-9]{4}-[0-9]{2}$/
          Time.parse(value, "%Y-%m", Time::Kind::Utc)
        when /^[0-9]{4}-[0-9]{2}-[0-9]{2}$/
          Time.parse(value, "%Y-%m-%d", Time::Kind::Utc)
        when /^[0-9]{4}-[0-9]{2}-[0-9]{2}([Tt]|[ \t]+)[0-9]{2}$/
          Time.parse(value.sub(/([Tt]|[ \t]+)/, 'T'), "%Y-%m-%dT%H", Time::Kind::Utc)
        when /^[0-9]{4}-[0-9]{2}-[0-9]{2}([Tt]|[ \t]+)[0-9]{2}:[0-9]{2}$/
          Time.parse(value.sub(/([Tt]|[ \t]+)/, 'T'), "%Y-%m-%dT%H:%M", Time::Kind::Utc)
        when /^[0-9]{4}-[0-9]{2}-[0-9]{2}([Tt]|[ \t]+)[0-9]{2}:[0-9]{2}:[0-9]{2}$/
          Time.parse(value.sub(/([Tt]|[ \t]+)/, 'T'), "%Y-%m-%dT%H:%M:%S", Time::Kind::Utc)
      {% end %}
      else
        raise YAML::ParseException.new("Expected #{name}, not #{value}", *location)
      end
    rescue
      raise YAML::ParseException.new("Expected #{name}, not #{value}", *location)
    end
  end
{% end %}
```